adr/feature-toggles/spike.md

```diff
@@ -0,0 +1,148 @@
+
+We have explored preview environments [[SCT-254]](https://hackney.atlassian.net/browse/SCT-254), and have reviewed our current ability to toggle features within our application [[SCT-399]](https://hackney.atlassian.net/browse/SCT-399).
+
+This document intends to investigate how we might implement a fairly lightweight feature toggle solution in our Next.js application, with the following goals:
+
+- Easy to implement new feature flags, with an obvious place for them to live
+- Easy to manage existing feature flags, and understand when they need to be tidied up
+- Transparency about which flags are currently in the systme, and how they are activated
+- That delivering a feature being a flag is so trivial that it becomes the default for all engineers on the team
+- That the API design has scope to grow to become more flexible in the future
+
+## How would a new solution work?
+
+The general concept is to introduce a single place for defining "feature toggles", which are made up of two properties, and an expiry date:
+
+```ts
+{
+  // FEATURE-FLAG-EXPIRES[2021-06-30] – a comment which notes when the feature toggle should be removed (this will cause an ESLint failure if the date is in the past)
+
+  name: 'some-feature'; // This is a unique identifier for a feature, that can be used later in code to reference it
+
+  get isActive(): boolean {
+    // A getter that returns a boolean, and determines if the flag is active or not.
+    //  Here you can check any of:
+    //  * `NEXT_PUBLIC_` environment variables
+    //  * Local/Session storage value (this should gracefully fail in a non-browser environment)
+    //  * Any other property you can think of! It's just TypeScript!
+
+    return true;
+  };
+}
+```
+
+> ⚠️ **Why is `isActive` a getter, not a function?** The same flag configuration is used on server render and client render. At render time, we:
+>
+> 1. Calculate the flag values on the server and pass them to the React application
+> 2. Render the app based on those flags at server render time
+> 3. Refresh the status of the flags on client hydration, and update them
+> 4. Optionally re-render the app if there are any flag changes (e.g. flags based on local storage values)
+>
+> The reason we have to use a [getter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) is that the feature flag object must be serialised so it can be passed between server and client, and we can't serialise a function. We _can_ however serialise a getter, because it consolidates down to it's return value, in our case, either `true` or `false`.
+
+Below is an example where we toggle a 'dashboard' feature based on two things (a Local Storage value or a environment variable). This feature is also set to 'expire' on June 30th 2021.
+
+```ts
+{
+  // FEATURE-FLAG-EXPIRES[2021-06-30]
+  name: 'dashboard',
+  get isActive() => {
+    if (getLocalStorageFlagValue('dashboard') === 'active') {
```

Comment:
Would a local storage flag value be something similar to values added to an `.env` file?

Reply:
Yeah the idea of using local storage is that if you want to quickly flip the value in your browser you could do. It was more of a demonstration of the capability than a suggestion we should actually use it. Depending on the feature flag you may find this useful, you may not. I do think we should build something like this using cookies though, so the value of the flag can be read on both the server and client, because it’s in a cookie!